src/core/message-parser.ts

```diff
@@ -10,48 +10,119 @@ interface MessagePayload {
+export function parseMessage(roomId: string | undefined, inboundMessage: Ably.InboundMessage): Message {
+  const message = inboundMessage as MessagePayload;
   if (!roomId) {
     throw new Ably.ErrorInfo(`received incoming message without roomId`, 50000, 500);
   }
 
-  if (!messageCreatedMessage.data) {
+  if (!message.data) {
     throw new Ably.ErrorInfo(`received incoming message without data`, 50000, 500);
   }
 
-  if (!messageCreatedMessage.clientId) {
+  if (!message.clientId) {
     throw new Ably.ErrorInfo(`received incoming message without clientId`, 50000, 500);
   }
 
-  if (!messageCreatedMessage.timestamp) {
+  if (!message.timestamp) {
     throw new Ably.ErrorInfo(`received incoming message without timestamp`, 50000, 500);
   }
 
-  if (messageCreatedMessage.data.text === undefined) {
+  if (message.data.text === undefined) {
     throw new Ably.ErrorInfo(`received incoming message without text`, 50000, 500);
   }
 
-  if (!messageCreatedMessage.extras) {
+  if (!message.extras) {
     throw new Ably.ErrorInfo(`received incoming message without extras`, 50000, 500);
   }
 
-  if (!messageCreatedMessage.extras.timeserial) {
+  if (!message.serial) {
     throw new Ably.ErrorInfo(`received incoming message without timeserial`, 50000, 500);
   }
 
-  return new DefaultMessage(
-    messageCreatedMessage.extras.timeserial,
-    messageCreatedMessage.clientId,
+  let operationDetails: MessageDetails | undefined;
+  if (message.operation) {
+    operationDetails = {
+      description: message.operation.description,
+      metadata: message.operation.metadata,
+    };
+  }
+
+  const newMessage: ChatMessageFields = {
+    timeserial: message.serial,
+    clientId: message.clientId,
     roomId,
-    messageCreatedMessage.data.text,
-    new Date(messageCreatedMessage.timestamp),
-    messageCreatedMessage.data.metadata ?? {},
-    messageCreatedMessage.extras.headers ?? {},
+    text: message.data.text,
+    createdAt: new Date(message.timestamp),
+    metadata: message.data.metadata ?? {},
+    headers: message.extras.headers ?? {},
+    updatedAt: message.updatedAt ? new Date(message.updatedAt) : undefined,
+    deletedAt: message.deletedAt ? new Date(message.deletedAt) : undefined,
+  };
+
+  switch (message.action) {
+    case 'MESSAGE_CREATE': {
+      break;
+    }
+    case 'MESSAGE_UPDATE': {
+      if (!message.updatedAt) {
+        throw new Ably.ErrorInfo(`received incoming update message without updatedAt`, 50000, 500);
+      }
+      newMessage.updatedBy = message.operation?.clientId;
+      newMessage.updateDetail = operationDetails;
+      break;
+    }
+    case 'MESSAGE_DELETE': {
+      if (!message.deletedAt) {
+        throw new Ably.ErrorInfo(`received incoming deletion message without deletedAt`, 50000, 500);
+      }
+      newMessage.deletedBy = message.operation?.clientId;
+      newMessage.deletionDetail = operationDetails;
+      break;
+    }
+    default: {
```

Comment:
I think we should move the `message.action` check to the top and simply ignore messages that don't have a known action.

There's been a push recently to make it so that SDKs do not complain if "new" fields are used - if we see an action that we don't understand, we should just ignore it (perhaps debug log it)

Reply:
I would tend to disagree, we are assuming by this stage that you are providing us with a message that fits the schema, if not, we should error. However, we have a check in messages to ignore messages with action types that we don't currently support.